### DISCUSSION TITLE
[24 Gennaio 2023] - Pipeline

### DISCUSSION THREAD

**Discussion by notedo** [CS-Swap/Architettura-degli-Elaboratori#24] [Pipeline]
### Traccia Esercizio

Il programma qui di seguito somma i primi L numeri nella sequenzafornita, limitando il valore massimo di ogni elemento a 4 (un elemento maggiore di 4 verrà limitato al valore 4). La stringa finale viene stampata se la somma totale è maggiore o uguale a 12.
Si supponga che tutte le istruzioni impiegate fanno parte del set supportato dalla CPU in figura, ossia non si fa uso di alcuna pseudoistruzione.
Si indichino (ignorando hazard che possano concernere la syscall):
1) le istruzioni tra le quali sono presenti data hazard – indicare i numeri di istruzione N ed M (visibili alla sinistra delle linee di codice in figura) ed il registro coinvolto $xY (nel formato N / M / $xY);
2) le istruzioni tra le quali sono presenti control hazard – indicare i numeri di istruzione N ed M (nel formato N / M);
3) quanti cicli di clock sono necessari ad eseguire il programma tramite forwarding, spiegando il calcolo effettuato;
4) quanti cicli di clock sarebbero necessari ad eseguire il programma senza forwarding, spiegando il calcolo effettuato;
5) quali sono, per ognuna delle cinque fasi, le istruzioni (o le bolle) in pipeline durante il 13° ciclo di clock (con forwarding);
6) qual è il ruolo delle connessioni al multiplexer a monte del PC nel diagramma seguente, spiegando nel maggior dettaglio possibile i tre possibili input e il ruolo dei due bit di controllo.

<img width="672" alt="img" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/110790224/cdb3aa59-55a4-4d06-a75b-4fa4d8ad0eba">

<img width="866" alt="img" src="https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/110790224/5f769aed-32d4-4c3c-b863-878aeaf21977">

**Comment by AglaiaNorza** [comment]
![image](https://github.com/CS-Swap/Architettura-degli-Elaboratori/assets/75022526/b1ce013f-eb48-4bf0-85b7-d588c16981bf)

Il MUX sopra il PC serve a spostare la decisione di jump alla fase IF.
I suoi segnali di controllo (e le loro utilità) sono:
1) un comparatore tra i bit che definiscono il tipo di operazione e l'opcode del jump (0000010=2) --> (riconosce se si tratta di un jump) se è 1 (caso 1x), viene passato al PC l'indirizzo a cui saltare, che sarà shiftato di 2 e concatenato a PC+4 per arrivare a 32 bit.
2) il segnale branchtaken --> se è 1 (caso 01), si sceglie PC+4 +etichetta del branch.
(i due segnali sono chiaramente mutualmente esclusivi).
 
Se entrambi sono 0, invece, passerà PC+4, senza nessun salto o branch.
